File: sources/com/qx/base/bytes/ByteInflow.java
package com.qx.base.bytes;

import java.io.IOException;


/**
 * <p>
 * Unified in-flow of bytes with getters for standard primitives.
 * </p>
 * 
 * <p>
 * /!\ All array or String lengths are encoded as UInt32 to reflect their
 * maximum possible length in JAVA.
 * </p>
 * 
 * @author pc
 *
 */
public interface ByteInflow {


	/**
	 * @return next byte
	 * @throws IOException 
	 */
	public byte getByte() throws IOException;


	/**
	 * length passed as argument
	 * 
	 * @param n next bytes, where n is the length of the <code>bytes</code> array
	 * @throws IOException 
	 */
	public byte[] getByteArray(int length) throws IOException;



	/**
	 * Tells if the next bytes in the ByteInput are matching the sequence passed as
	 * argument. Note that this method will return false if there is not enough
	 * bytes to make the comparison.
	 * 
	 * @param bytes
	 * @return
	 * @throws IOException
	 */
	public boolean isMatching(byte[] sequence) throws IOException;


	/**
	 * @return an int decoded with the multi-bytes format.
	 * 
	 * @throws IOException
	 */
	public int getUInt() throws IOException;


	/**
	 * @return the next Unsigned Integer 8 bits (1 byte)
	 * @throws IOException
	 */
	public int getUInt8() throws IOException;


	/**
	 * boolean
	 * @return
	 * @throws IOException
	 */
	boolean[] getFlagsBlock() throws IOException;


	/**
	 * @return the next Unsigned Integer 16 bits (2 bytes)
	 * @throws IOException
	 */
	public int getUInt16() throws IOException;


	/**
	 * @return the next Signed Integer 16 bits (2 bytes)
	 * @throws IOException
	 */
	public short getInt16() throws IOException;


	/**
	 * @return the next Unsigned Integer 31 bits (4 bytes, first bit ignored)
	 * @throws IOException
	 */
	public int getUInt31() throws IOException;



	/**
	 * Curretly doing the same as <code>getUInt31</code>
	 * @return
	 * @throws IOException
	 */
	public int getUInt32() throws IOException;

	/**
	 * @return the next Signed Integer 32 bits (4 bytes)
	 * @throws IOException
	 */
	public int getInt32() throws IOException;


	/**
	 * Number of elements is encoded as an UInt16 before starting the first value.
	 * The max number of elements is max(UInt16) = 2^16 = 65536.
	 * 
	 * @return the next Signed Integer 32 bits Array (4 bytes for each value).
	 * @throws IOException
	 */
	public int[] getInt32Array() throws IOException;

	/**
	 * @return the next Signed Integer 64 bits (8 bytes)
	 * @throws IOException
	 */
	public long getInt64() throws IOException;


	/**
	 * Number of elements is encoded as an UInt16 before starting the first value.
	 * The max number of elements is max(UInt16) = 2^16 = 65536.
	 * 
	 * @return the next Signed Integer 64 bits Array (8 bytes for each value).
	 * @throws IOException
	 */
	public long[] getInt64Array() throws IOException;


	/**
	 * @return the next Float 32 bits (4 bytes)
	 * @throws IOException
	 */
	public float getFloat32() throws IOException;


	/**
	 * Number of elements is encoded as an UInt16 before starting the first value.
	 * The max number of elements is max(UInt16) = 2^16 = 65536.
	 * 
	 * @return the next Float 32 bits Array (4 bytes for each value).
	 * @throws IOException
	 */
	public float[] getFloat32Array() throws IOException;


	/**
	 * @return the next Float 64 bits (8 bytes)
	 * @throws IOException
	 */
	public double getFloat64() throws IOException;


	/**
	 * Number of elements is encoded as an UInt16 before starting the first value.
	 * The max number of elements is max(UInt16) = 2^16 = 65536.
	 * 
	 * @return the next Float 64 bits Array (8 bytes for each value).
	 * @throws IOException
	 */
	public double[] getFloat64Array() throws IOException;


	/**
	 * String is encoded in UTF-8.
	 * 
	 * @return the next String.
	 * @throws IOException
	 */
	public String getString() throws IOException;



}

File: web-helium/ByteInflow.js


/**
 * <p>
 * Unified in-flow of bytes with getters for standard primitives.
 * </p>
 * 
 * <p>
 * /!\ All array or String lengths are encoded as UInt32 to reflect their
 * maximum possible length in JAVA.
 * </p>
 * 
 * @author pc
 *
 */
export class ByteInflow {

	/**
	 * 
	 * @param {ArrayBuffer} arraybuffer 
	 */
	constructor(arraybuffer) {
		this.arraybuffer = arraybuffer;
		this.view = new DataView(arraybuffer);
		this.textDecoder_ASCII = new TextDecoder('ascii');
		this.textDecoder_UTF8 = new TextDecoder('utf8');
		this.offset = 0;
	}


	/**
	 * 
	 * @returns 
	 */
	getByte() {
		let value = this.view.getUint16(this.offset);
		this.offset += 1;
		return value;
	}


	/**
	 * 
	 * @returns 
	 */
	getUInt8() {
		let value = this.view.getUint8(this.offset);
		this.offset += 1;
		return value;
	}


	/**
	 * 
	 * @returns 
	 */
	getUInt16() {
		let value = this.view.getUint16(this.offset);
		this.offset += 2;
		return value;
	}


	/**
	 * 
	 * @returns 
	 */
	getInt16() {
		let value = this.view.getInt16(this.offset);
		this.offset += 2;
		return value;
	}

	getUInt32() {
		let value = this.view.getUint32(this.offset);
		this.offset += 4;
		return value;
	}

	getInt32() {
		let value = this.view.getInt32(this.offset);
		this.offset += 4;
		return value;
	}

	getUInt64() {
		let value = this.view.getUInt64(this.offset);
		this.offset += 8;
		return value;
	}

	getInt64() {
		let value = this.view.getInt64(this.offset);
		this.offset += 8;
		return value;
	}


	getFloat32() {
		let value = this.view.getFloat32(this.offset);
		this.offset += 4;
		return value;
	}

	getFloat64() {
		let value = this.view.getFloat64(this.offset);
		this.offset += 8;
		return value;
	}

	getL8StringASCII() {
		let length = this.getUInt8();
		let stringView = new Uint8Array(this.arraybuffer, this.offset, length);
		let value = this.textDecoder_ASCII.decode(stringView);
		this.offset += length;
		return value;
	}

	getL32StringUTF8() {
		let length = this.getUInt32();
		let stringView = new Uint8Array(this.arraybuffer, this.offset, length);
		let value = this.textDecoder_UTF8.decode(stringView);
		this.offset += length;
		return value;
	}


	getIndex(length){
		let script = "";
		for(let i=0; i<length; i++){
			let digit = this.view.getUint8(this.offset);
			this.offset += 1;
			script += digit > 0x0f ? digit.toString(16) : '0'+digit.toString(16);
		}
		return script;
    }

}